cpp/src/arrow/compute/kernels/take.cc

```diff
@@ -0,0 +1,221 @@
+    IndexArrayRef indices = static_cast<IndexArrayRef>(*params_.indices);
+    ValueArrayRef values = static_cast<ValueArrayRef>(*params_.values);
+    std::unique_ptr<ArrayBuilder> builder;
+    RETURN_NOT_OK(MakeBuilder(params_.context->memory_pool(), values.type(), &builder));
+    RETURN_NOT_OK(builder->Reserve(indices.length()));
+    RETURN_NOT_OK(TakeImpl(params_.context, values, indices, params_.options,
+                           static_cast<OutBuilder*>(builder.get())));
+    return builder->Finish(params_.out);
+  }
+
+  Status Visit(const NullType& t) {
+    auto indices_length = params_.indices->length();
+    if (params_.options.out_of_bounds == TakeOptions::ERROR && indices_length != 0) {
+      auto indices = static_cast<IndexArrayRef>(*params_.indices).raw_values();
+      auto max = *std::max_element(indices, indices + indices_length);
+      if (static_cast<int64_t>(max) > params_.values->length()) {
+        return Status::Invalid("out of bounds index");
+      }
+    }
+    params_.out->reset(new NullArray(indices_length));
+    return Status::OK();
+  }
+
+  Status Visit(const DictionaryType& t) {
+    auto dictionary_indices = params_.values->data()->Copy();
+    dictionary_indices->type = t.index_type();
+    TakeParameters params = params_;
+    params.values = MakeArray(dictionary_indices);
+    UnpackValues<IndexType> unpack = {params};
+    RETURN_NOT_OK(VisitTypeInline(*t.index_type(), &unpack));
+    (*params_.out)->data()->type = dictionary(t.index_type(), t.dictionary());
+    return Status::OK();
+  }
+
+  Status Visit(const ExtensionType& t) {
+    // XXX can we just take from its storage?
+    return Status::NotImplemented("gathering values of type ", t);
+  }
+
+  Status Visit(const UnionType& t) {
+    return Status::NotImplemented("gathering values of type ", t);
+  }
+
+  Status Visit(const ListType& t) {
+    return Status::NotImplemented("gathering values of type ", t);
+  }
+
+  Status Visit(const StructType& t) {
+    return Status::NotImplemented("gathering values of type ", t);
+  }
+
+  const TakeParameters& params_;
+};
+
+struct UnpackIndices {
+  template <typename IndexType>
+  enable_if_integer<IndexType, Status> Visit(const IndexType&) {
+    UnpackValues<IndexType> unpack = {params_};
+    return VisitTypeInline(*params_.values->type(), &unpack);
+  }
+  Status Visit(const DataType& other) {
+    return Status::Invalid("index type not supported: ", other);
+  }
+  const TakeParameters& params_;
+};
+
+Status TakeKernel::Call(FunctionContext* ctx, const Datum& values, const Datum& indices,
+                        Datum* out) {
+  if (!values.is_array() || !indices.is_array()) {
+    return Status::Invalid("TakeKernel expects array values and indices");
+  }
+  std::shared_ptr<Array> out_array;
+  TakeParameters params;
+  params.context = ctx;
+  params.values = values.make_array();
+  params.indices = indices.make_array();
+  params.options = options_;
+  params.out = &out_array;
```

Comment:
Why is it a reference, can it be the actual ptr?

Reply:
Mostly I was following the convention that output arguments are passed by pointer.

Since I only make one instance of TakeParameters and pass it around by `const&` its data members are not mutable. I can refactor if you think it's better to:
- mark `out` mutable
- make copies of `TakeParameter`
- pass `TakeParameter` around by mutable reference